### PR TITLE
payload info

### PR DESCRIPTION
**Pull request by Slava0135**
Based off Quezler's work
Screenshot from 2020-12-26 15-20-56

### CODE REVIEW

**Comment by Slava0135** [comment]
I want to make player be able to look what his unit is carrying on but idk how should I do this

**Comment by Quezler** [comment]
interesting how you stacked them half over the previous one 👍

File: core/src/mindustry/type/UnitType.java [hunk 1]
```diff
@@ -171,6 +171,27 @@ public void display(Unit unit, Table table){
...
+
+            if(unit instanceof Payloadc){
+                bars.add(new Bar("stat.payloadcapacity", Pal.items, () -> ((Payloadc)unit).payloadUsed() / unit.type().payloadCapacity));
+                bars.row();
+
+                bars.table(req -> {
+                    req.top().left();
+
+                    float pad = 1;
+                    float items = ((Payloadc)unit).payloads().size;
```
Comment (summetdev) [inline core/src/mindustry/type/UnitType.java:183]: use pattern matching instead of cast:
```
if(unit instanceof Payloadc p){
    p.payloads();
    // ...
}
```

**Comment by Slava0135** [comment]
Effect changes and show player unit info when nothing has been caught
Peek 2020-12-26 16-48

**Comment by Slava0135** [comment]
however as you can see it doesn't update the table...

File: core/src/mindustry/entities/comp/PayloadComp.java [hunk 2]
```diff
@@ -61,7 +61,7 @@ void pickup(Building tile){
         tile.pickedUp();
         tile.tile.remove();
         payloads.add(new BuildPayload(tile));
-        Fx.unitPickup.at(tile);
+        Fx.breakBlock.at(tile, tile.block.size);
```
Comment (Anuken) [inline core/src/mindustry/entities/comp/PayloadComp.java:64]: Why did you change this? The block isn't being broken; this effect is misleading and doesn't look good.

File: core/src/mindustry/type/UnitType.java [hunk 3]
```diff
@@ -171,6 +171,27 @@ public void display(Unit unit, Table table){
...
+                    req.top().left();
+
+                    float pad = 1;
+                    float items = payload.payloads().size;
+                    if((8 * 2) * items + pad * items > 270f){
+                        pad = (270f - (8 * 2) * items) / items;
+                    }
+
+                    float finalPad = pad;
+                    payload.payloads().each(p -> {
```
Comment (Anuken) [inline core/src/mindustry/type/UnitType.java:189]: Don't use `each` here, use a normal loop. Final variables are unnecessary.

File: core/src/mindustry/ui/fragments/PlacementFragment.java [hunk 4]
```diff
@@ -469,6 +469,6 @@ Displayable hovered(){
             }
         }
 
-        return null;
+        return player.unit();
```
Comment (Anuken) [inline core/src/mindustry/ui/fragments/PlacementFragment.java:472]: Instead of hovering, the player status at the top left should display payloads.

File: core/src/mindustry/type/UnitType.java [hunk 5]
```diff
@@ -171,6 +171,27 @@ public void display(Unit unit, Table table){
...
+
+            if(unit instanceof Payloadc payload){
+                bars.add(new Bar("stat.payloadcapacity", Pal.items, () -> payload.payloadUsed() / unit.type().payloadCapacity));
+                bars.row();
+
+                bars.table(req -> {
+                    req.top().left();
+
+                    float pad = 1;
+                    float items = payload.payloads().size;
```
Comment (Anuken) [inline core/src/mindustry/type/UnitType.java:183]: The table should be rebuilt whenever the payload count changes. This isn't straightforward to implement with Java's `effectively final` system, but what you need to do is store the count in a 1-len array and run a local runnable to rebuild when it changes.

**Review by Anuken — CHANGES_REQUESTED** [review]

File: core/src/mindustry/entities/comp/PayloadComp.java (hunk 2 above)
Comment (Slava0135) [reply core/src/mindustry/entities/comp/PayloadComp.java:64]: Because it's hard to see whether you picked it up or not. However with new ui it's more obvious.

**Comment by Slava0135** [comment]
Peek 2020-12-26 21-21
so now it updates (similar to item requirements table)

File: core/src/mindustry/type/UnitType.java [hunk 6]
```diff
@@ -171,6 +172,16 @@ public void display(Unit unit, Table table){
...
                 bars.row();
             }
+
+            if(unit instanceof Payloadc payload){
+                bars.add(new Bar("stat.payloadcapacity", Pal.items, () -> payload.payloadUsed() / unit.type().payloadCapacity));
+                bars.row();
+
+                bars.add(new Table()).update(it -> {
+                    payload.contentInfo(it, 8 * 2, 270);
+                }).growX().left();
```
Comment (summetdev) [inline core/src/mindustry/type/UnitType.java:182]:
```suggestion
                bars.table().update(t -> {
                    payload.contentInfo(t, 8 * 2, 270);
                }).growX().left();
```

**Comment by Slava0135** [comment]
@Anuken I was trying my best to make gui work but it just doesn't work.
the idea was (by Ilya246):
image
image
^just put the table below this

**Comment by Slava0135** [comment]
h O w T o F i T t H e M
https://user-images.githubusercontent.com/40753025/103167992-db57c000-4840-11eb-9be3-5a0970af59f3.mp4

image

**Comment by Slava0135** [comment]
image

https://user-images.githubusercontent.com/40753025/103168172-0b539300-4842-11eb-8721-f60666dcbe6e.mp4

**Comment by Slava0135** [comment]
https://user-images.githubusercontent.com/40753025/103168251-8fa61600-4842-11eb-8e88-902a9efe0ca6.mp4

**Comment by Slava0135** [comment]
Screenshot from 2020-12-27 13-07-26
image

**Comment by Slava0135** [comment]
Squashed

**Comment by Slava0135** [comment]
Uploading aaaaaaaaaa.mp4…
image
*we have problems...*

**Comment by Slava0135** [comment]
Screenshot from 2020-12-27 13-07-26
image

**Comment by Slava0135** [comment]
https://user-images.githubusercontent.com/40753025/103203482-11ab4300-4906-11eb-8760-9540731c7f90.mp4

**Comment by Slava0135** [comment]
and we have another problem...
image

**Comment by Slava0135** [comment]
https://user-images.githubusercontent.com/40753025/103204572-b2026700-4908-11eb-9b39-2965ebd51e7b.mp4

File: core/src/mindustry/ui/fragments/HudFragment.java [hunk 7]
```diff
@@ -755,6 +806,16 @@ public void draw(){
             return builder;
         }).growX().pad(8f);
 
+        table.row();
+
+        table.table().update(t -> {
+            if(player.unit() instanceof Payloadc payload){
+                payload.contentInfo(t, 8 * 2, 280f);
```
Comment (Anuken) [inline core/src/mindustry/ui/fragments/HudFragment.java:813]: Updating the payload info **every frame** will lead to a lot of extra garbage. This should only happen when payloads are picked up/dropped.

**Comment by Slava0135** [comment]
reverted bars

**Comment by Anuken** [comment]
Implemented in linked commit. May add bars for capacity later.
